draft-ietf-quic-multipath.md

```diff
@@ -625,6 +625,26 @@ delay will most of the time relate to the path with the shortest latency.
 To collect ACK delays on all the paths, hosts should rely on time stamps
 as described in {{QUIC-Timestamp}}.
 
+### Handling ECN Feedback
+
+ECN feedback in QUIC is provided based on counters in the ACK frame
+(see {{Section  19.3.2. of QUIC-TRANSPORT}}). That means if an ACK
+frame acknowledges multiple packets, the ECN feedback cannot be accounted
+to a specific packet. Respectively, if an ACK frames acknowledges multiple
+packets from different paths, the ECN feedback cannot unambiguously be assigned to
+a path. Therefore if one of the ECN counters increases, especially the CE
+counter, hosts SHOULD only acknowlegde packets from the same path where
+the ECN mark was observed in the same ACK frame.
+
```

Comment:
I would change to:

Respectively, if an ACK frames acknowledges multiple
packets from different paths, the ECN counters provide the sum of all signals received on those paths.

Reply:
not sure if that is better, because the problem is that you don't know anymore to which path the CE belong when you received the ACK. I think your text proposal makes that point less clear.

Reply:
I think the acknowledgement process should be completely straightforward:

* when ECN mark received, increment the counters for the number space of the packet
* when sending ACK ECN, send the ECN for the number space of the ACK

Yes, in the case of single number space, that does mean single counter for all paths, and that makes the signal harder to use in congestion control. Implementers should be warned about that. But notice that the ACK-ECN are sent by the receiver, and the congestion control decisions are made by the sender. The sender does not have the option of "not receiving ACK-ECN" -- there is no parameter negotiation for that. The sender has different options:

* just not use ECN, i.e., send all packets with ECN flags "00".
* use ECN on only some path, e.g., the one path on which most data is sent, mark packets on this path with ECT-0 or ECT-1, mark packets on the other paths as "00".
* use ECT-0 on one path, ECT-1 on the other, and use some logic to attribute ECN counts to either ECT-0 or ECT-1.
* use ECT-0 or ECT-1 on all paths, and use correlation with other factors (e.g., losses, delays) to allocate the ECN marks to different paths.

I would not preclude any of these options.

Reply:
I think it would be better to discuss this in the issue #87 and not this PR but let me reply directly to a couple of points from you:

First, as you note this problem only exists for the single PNS solution. Therefore please note that this PR proposes to add this text to the respective section for single PNS considerations. We can add more text about general ECN but that shouldn't be much different from normal QUIC operation and would be a separate PR for a different section of the document.  

> Yes, in the case of single number space, that does mean single counter for all paths, and that makes the signal harder to use in congestion control.

I don't think "harder to use" is right. If multiple packets from different paths are ack'ed in the same frame it's simply impossible to assign the signal to one path. As you never should just ignore a congestion signal, the most conservative but probably only correct reaction is unfortunately to reduce your cwnd on all paths where packet got ack'ed. 

> just not use ECN, i.e., send all packets with ECN flags "00".

Yes, that is an option and I can add this. This was actually my first thought (see my reply in issue #87). However, for me that would be a strong argument against the single PNS solution. But then I thought that if the receiver would send separate ACKs per path that could also address the problem and keeping ACKs separated per path (at least if the sender sends blocks of PNs on each path) might be beneficial anyway. I think that also something that Ian had in mind.

> use ECN on only some path, e.g., the one path on which most data is sent, mark packets on this path with ECT-0 or ECT-1, mark packets on the other paths as "00".

That means you could only see ECN on one path. I don't think this an appropriate solution and again would be reason for me to not go for the single PNS solution.

> use ECT-0 on one path, ECT-1 on the other, and use some logic to attribute ECN counts to either ECT-0 or ECT-1.

I think this is out of scope for this document as it makes a different use of ECT1 against the recommendation in RFC8311 (and in conflict with L4S). So if you want to propose this, you have to go to tsvwg.

> use ECT-0 or ECT-1 on all paths, and use correlation with other factors (e.g., losses, delays) to allocate the ECN marks to different paths.

I don't think this could work. You don't have per packet delay information in the ACK, so that exactly the same ambiguity. And ideally loss should not be correlated with ECN - hopefully you have ECN instead of loss but that also not guaranteed as you might have multiple bottlenecks or other reason for loss.